### PR TITLE
make complement, reverse_complement consistent

### PR DESCRIPTION
**Pull request by mdehoon**
Currently, the `complement` and `reverse_complement` methods and functions in `Bio.Seq` are inconsistent in their handling of RNA sequences and mutable sequences. Sometimes the presence of a `U` indicates RNA, sometimes the presence of a `u` or `U` means RNA, sometimes mutable sequences are acted on in place, sometimes a `Seq` object is returned.

This PR aims to make `complement` and `reverse_complement` consistent by adding an `inplace` keyword argument (same as for `strip`, `upper` etc. in the same module).

**In the future**, `inplace` will be `False` by default, and `inplace=None` will be disallowed.
**For now**, `inplace` is `None` by default.

- If `inplace` is `False`, the function/method does not act in place, and returns a new object of the same type as the sequence it acts on;
- If `inplace` is `True`, the function/method tries to act in place, and raises a `TypeError` if the sequence passed is immutable. If successful, the function/method returns the sequence is acted on;
- If `inplace` is `True` or `False`, then
    - `complement` and `reverse_complement` return a DNA sequence;
    - `complement_rna` and `reverse_complement_rna` return an RNA sequence.
- If `inplace` is `None`, then `complement`, `reverse_complement` will act the same as in release <= 1.79.  However, if function/method will behave differently for the sequence provided once the default value of `inplace` becomes `False`, a `BiopythonDeprecationWarning` will be issued encouraging the user to either set `inplace=True`, or use `complement_rna` and `reverse_complement_rna`.

In a few places, `inplace=False` was added to calls to `complement`, `reverse_complement` to avoid the deprecation warning; these can be removed once the default value of `inplace` becomes `False`.

The tutorial documentation has not yet been fully updated; I wanted to see first if there are any strong objections to this PR.

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit`` locally,
and understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### CODE REVIEW

**Comment by mdehoon** [comment]
@peterjc  Any objections?

This PR also includes a workaround for the PyPy bug we saw with numpy.

**Comment by mdehoon** [comment]
Ping @peterjc

**Comment by mdehoon** [comment]
@peterjc  I'd like to merge this before moving on to #3667 .If you have no objections, I'll proceed; in any case we can still makes changes to this code, as release 1.80 is probably still far away.

File: Bio/SeqUtils/__init__.py [hunk 1]
```diff
@@ -380,13 +380,16 @@ def molecular_weight(
             "%s is not a valid unambiguous letter for %s" % (e, seq_type)
         ) from None
 
-    if seq_type in ("DNA", "RNA") and double_stranded:
-        seq = str(Seq(seq).complement())
+    if double_stranded:
+        if seq_type == "protein":
+            raise ValueError("double-stranded proteins await their discovery")
```
Comment (peterjc) [inline Bio/SeqUtils/__init__.py:385]: Funny, but perhaps a clearer error message is wiser?

Comment (peterjc) [reply Bio/SeqUtils/__init__.py:385]: Ah, this was the message - just the code moved. Oh well, leave it as it.

File: Bio/motifs/__init__.py [hunk 2]
```diff
@@ -237,10 +237,17 @@ def search(self, sequence):
...
+
         instances = Instances(alphabet=self.alphabet)
         instances.length = self.length
         for instance in self:
-            instance = instance.reverse_complement()
+            # TODO: remove inplace=False
+            if isinstance(instance, (Seq, MutableSeq)):
+                instance = instance.reverse_complement(inplace=False)
+            if isinstance(instance, (str, SeqRecord)):
+                instance = instance.reverse_complement()
```
Comment (peterjc) [inline Bio/motifs/__init__.py:250]: This looks wrong - what if the instance is none of the above? You won't attempt to take a reverse complement, nor raise an exception.

**Review by peterjc — CHANGES_REQUESTED** [review]
There looks to be an oversight in the motifs ``reverse_complement``, otherwise seems OK.

File: Bio/motifs/__init__.py (hunk 2 above)
Comment (mdehoon) [reply Bio/motifs/__init__.py:250]: Thanks. I agree that it's better to raise an exception here if it is none of the above. The instances were already checked when the object was created, but a user might have added more instances afterwards.

**Pull request merged**